src/csharp/Microsoft.Spark/Sql/StorageLevel.cs

```diff
@@ -0,0 +1,242 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+
+namespace Microsoft.Spark.Sql
+{
+    /// <summary>
+    /// Flags for controlling the storage of an RDD. Each StorageLevel records whether to use
+    /// memory, whether to drop the RDD to disk if it falls out of memory, whether to keep the
+    /// data in memory in a JAVA-specific serialized format, and whether to replicate the RDD
+    /// partitions on multiple nodes. Also contains static properties for some commonly used
+    /// storage levels, MEMORY_ONLY.
+    /// </summary>
+    public sealed class StorageLevel : IJvmObjectReferenceProvider
+    {
+        private static readonly string s_storageLevelClassName =
+            "org.apache.spark.storage.StorageLevel";
+        private static StorageLevel s_none;
+        private static StorageLevel s_diskOnly;
+        private static StorageLevel s_diskOnly2;
+        private static StorageLevel s_memoryOnly;
+        private static StorageLevel s_memoryOnly2;
+        private static StorageLevel s_memoryOnlySer;
+        private static StorageLevel s_memoryOnlySer2;
+        private static StorageLevel s_memoryAndDisk;
+        private static StorageLevel s_memoryAndDisk2;
+        private static StorageLevel s_memoryAndDiskSer;
+        private static StorageLevel s_memoryAndDiskSer2;
+        private static StorageLevel s_offHeap;
+        private readonly JvmObjectReference _jvmObject;
+
+        JvmObjectReference IJvmObjectReferenceProvider.Reference => _jvmObject;
+
+        internal StorageLevel(JvmObjectReference jvmObject)
+        {
+            UseDisk = (bool)jvmObject.Invoke("useDisk");
+            UseMemory = (bool)jvmObject.Invoke("useMemory");
+            UseOffHeap = (bool)jvmObject.Invoke("useOffHeap");
+            Deserialized = (bool)jvmObject.Invoke("deserialized");
+            Replication = (int)jvmObject.Invoke("replication");
+            _jvmObject = jvmObject;
+        }
+
+        public StorageLevel(
+            bool useDisk,
+            bool useMemory,
+            bool useOffHeap,
+            bool deserialized,
+            int replication = 1)
+        {
+            UseDisk = useDisk;
+            UseMemory = useMemory;
+            UseOffHeap = useOffHeap;
+            Deserialized = deserialized;
+            Replication = replication;
+            _jvmObject = (JvmObjectReference)SparkEnvironment.JvmBridge.CallStaticJavaMethod(
```

Comment:
Changed, however I found I still need to set `_jvmObject` in the second constructor, otherwise it remains null after the `StorageLevel` object is created, despite going through the first constructor and assigning `_jvmObject = jvmObject`.

Reply:
oh really? even doing `:this(...)`?

Reply:
that worked, thanks!